flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultAllocatedSlotPool.java

```diff
@@ -234,4 +241,30 @@ private static DefaultFreeSlotInfo create(
                     Preconditions.checkNotNull(slotInfoWithUtilization), idleSince);
         }
     }
+
+    private static final class DefaultAllocatedSlotsAndReservationStatus
+            implements AllocatedSlotsAndReservationStatus {
+
+        private final Map<AllocatedSlot, ReservationStatus> slotsAndReservationStatus;
```

Comment:
`AllocatedSlot` is not intended to be a key for hash maps. Hence, it does not implement a proper `hashCode` method. I would suggest to use the `AllocationID`.